Real-time collaboration: Use relative positions in undo stack

alecgeatches opened the pull request:
## What?

Improve undo positioning in real-time collaborative documents.

Our current implementation of the `UndoStack` with real-time collaboration [undoes stack operations](https://github.com/WordPress/gutenberg/blob/9714d0a9bd7fcaeee3e1b80514e5854ccddb5ef3/packages/sync/src/undo-manager.ts#L68-L69) but does not manage the user's cursor position within the undo stack. This leads to a lot of unexpected jumps, selection loss, or relocation to the beginning of the line when changes are undone.

*Undo in the current `wpvip/rtc-plugin` branch*:

https://github.com/user-attachments/assets/1689bf57-b139-4dd2-a6a0-adec191cb874

The user's cursor can jump to an unexpected place or lose focus when an undo operation happens. This is caused from restoring content but not the user's cursor.

The changes in this PR use [`Y.UndoManager`](https://docs.yjs.dev/api/undo-manager#example-add-additional-information-to-the-stackitems)'s `stack-item-added` and `stack-item-popped` events to store the user's [`RelativePosition`](https://docs.yjs.dev/api/relative-positions) on the stack. This fixes regular undo behavior, and also accounts for position changes that might have been made by another user in the CRDT doc. Here are the same tests above in this branch:

*Undo in this PR (`fix/relative-position-undo`)*:

https://github.com/user-attachments/assets/b7e3d0b3-12b0-4633-970d-48696ef91b30

## Why?

Because our real-time undo stack relies on `Y.UndoManager`, we need to give it information on the user's cursor when we save items to the stack, or change the user's cursor when we pop items off. We store [`RelativePosition`](https://docs.yjs.dev/api/relative-positions) objects instead of absolute offsets which allow cursor positions to move if the underlying content has changed.

## How?

This PR relies on a couple of important pieces:

- Viewing the user's current selection, and restoring selection in the `sync`'s `UndoManager` is difficult without introducing circular dependencies. To work around this, we pass `addUndoMeta()` and `restoreUndoMeta()` record handlers to core data, which the sync manager uses to pass undo stack meta and activate it during an undo operation.
- The `BlockSelectionHistory` lives in core data class keeps a recent set of selected blocks, bucketed by start and end block clientId, which was found in practice to be a good way to store recent selections without duplicated data. For example, here's a user typing in two blocks:

    ![b1-b2](https://github.com/user-attachments/assets/e3e8df5a-5847-4e31-8f1b-326dccfb9eed)

    The de-duplicated selection updates from the block editor store give these set of selection updates:

    ```js
    { "selectionStart": {}, "selectionEnd": {} }
    {
        "selectionStart": { "clientId": "<block1>", "attributeKey": "content", "offset": 0 },
        "selectionEnd":   { "clientId": "<block1>", "attributeKey": "content", "offset": 0 }
    }
    {
        "selectionStart": { "clientId": "<block1>", "attributeKey": "content", "offset": 1 },
        "selectionEnd":   { "clientId": "<block1>", "attributeKey": "content", "offset": 1 }
    }
    {
        "selectionStart": { "clientId": "<block1>", "attributeKey": "content", "offset": 2 },
        "selectionEnd":   { "clientId": "<block1>", "attributeKey": "content", "offset": 2 }
    }
    {
        "selectionStart": { "clientId": "<block2>" },
        "selectionEnd":   { "clientId": "<block2>" }
    }
    {
        "selectionStart": { "clientId": "<block2>", "attributeKey": "content" },
        "selectionEnd":   { "clientId": "<block2>", "attributeKey": "content" }
    }
    {
        "selectionStart": { "clientId": "<block2>", "attributeKey": "content", "offset": 0 },
        "selectionEnd":   { "clientId": "<block2>", "attributeKey": "content", "offset": 0 }
    }
    {
        "selectionStart": { "clientId": "<block2>", "attributeKey": "content", "offset": 1 },
        "selectionEnd":   { "clientId": "<block2>", "attributeKey": "content", "offset": 1 }
    }
    {
        "selectionStart": { "clientId": "<block2>", "attributeKey": "content", "offset": 2 },
        "selectionEnd":   { "clientId": "<block2>", "attributeKey": "content", "offset": 2 }
    }
    ```

    `<block1>` and `<block2>` are shortened block clientIds.

    There's a lot of noise in here, especially when blocks are created like `<block2>` which send an update with just the`clientId`, then the `clientId` and `attributeKey`, and finally the location of the cursor when the block is done being created with an `offset`.  It can be tricky to determine which of these represent a valid state for the user's selection to return to.

    `BlockSelectionHistory` processes these updates to bucket them by `clientId`. At the end of the actions above, here's what `BlockSelectionHistory` has stored:

    ```js
    {
        "currentSelection": {
            "start": {
                "type": "RelativeSelection",
                "attributeKey": "content",
                "relativePosition": { /* ... */ },
                "clientId":"<block2>",
                "offset":2
            },
            "end": { /** same as start */ },
        }
        "history": [
            {
                "start": {
                    "type": "RelativeSelection",
                    "attributeKey": "content",
                    "relativePosition": { /* ... */ },
                    "clientId":"<block1>",
                    "offset":2
                },
                "end": { /** same as start */ },
            }
        ]
    }
    ```

    where the `relativePosition` property is a [`Y.RelativePosition`](https://docs.yjs.dev/api/relative-positions) type.

    Now when we process an undo stack being added, it's easy to grab the current relative position and some recent backups and push them onto the stack. The current implementation saves the current selection along with the last three unique block selections made before it.
    
    When the user undoes an item in the undo stack, we [iterate through the positions](https://github.com/WordPress/gutenberg/blob/b3d75ad10705113c75f7979e7de36ad1910e3ef4/packages/core-data/src/utils/crdt.ts#L548) on that item and move to the first valid one. In testing, especially in a multi-user environment, that was found to be a robust way to handle fallback positions when the underlying document can change.

### Selection ranges

We also store the `start` and `end` for each selection, which allows us to restore selected content made before a change:

![restore-selected-content](https://github.com/user-attachments/assets/3033d316-698e-43c4-8c8a-78c2a9ff2e99)

This is change from non-collaborative Gutenberg, which does not keep selection state in the undo stack.

## What's missing?

There are still some features we can enhance to improve the undo stack or match WordPress behavior better:

### 1. Match WordPress behavior by creating an undo stack item when a block is inserted

WordPress will create a new undo location at the beginning of each block:

![wordpress-line-undo](https://github.com/user-attachments/assets/cba1d98b-caf1-4f2f-a454-2a7c98bc4829)

When undoing, the cursor stops at each new block. However, `Y.UndoManager` instead uses a [`captureTimeout` of `500ms`](https://github.com/WordPress/gutenberg/blob/2449f9ef0ee549efbb490adee47c125bf1103f79/packages/sync/src/undo-manager.ts#L53) to decide when new stack items are added. If typed quickly enough, all 3 blocks are grouped together in one stack item instead:

![rtc-line-undo](https://github.com/user-attachments/assets/7975736f-4378-4943-81bb-095b838961ea)

This may be possible to fix, but in the current behavior `Y.UndoManager` is fully in charge of when stack items are committed and we don't have specific say other than `captureTimeout`.

### 2. Block lookup inefficiencies

In order to convert a cursor offset to a `RelativePosition`, we need access to the underlying `Y.Text` object that represents the block's RichText content. This currently relies on the [`findBlockByClientIdInDoc()` function](https://github.com/WordPress/gutenberg/blob/2449f9ef0ee549efbb490adee47c125bf1103f79/packages/sync/src/utils.ts#L95), which recursively scans a document's entire content to find the associated `Y.Text` object. This works fine for small document tests but may be a bottleneck in larger documents.

This is the same inefficient approach we use within the [vip-real-time-collaboration plugin](https://github.com/Automattic/vip-real-time-collaboration/blob/9a378d4834af6812d33d078b0b399a8701f44ae3/src/utilities/selection.ts#L226-L229), but it should be changed in both to a better system. I think a top level map of `clientId -> attributeName -> Y.Text` would be a much more efficient way to store `Y.Text` values for lookup and would solve the O(N) runtime of the current implementation, but this will require changes to attribute storage.

## Testing Instructions

1. Check out this PR.
2. As a single user, type things in the editor, create new blocks, and move the cursor around. Try undoing and redoing actions, an ensure that the cursor position returns to an intuitive position.
3. Try selecting a range within a single block, type over the selected content, and then undo. The selection should be restored. Note that this behavior doesn't work when restoring a selection across multiple blocks, which is a limitation of selection state in Gutenberg.
4. Open another tab with another user and edit the document at the same time. Ensure that undo and redo operations continue to work, even if blocks are changed. Since we store a `RelativePosition`, the user's cursor should still return to the correct logical spot in a previous paragraph even if the offset has changed.

There are also a set of tests in `block-selection-history.test.ts`:

```
npm run test:unit -- --testPathPattern=block-selection-history
```

alecgeatches committed:
Track prior user relative position in UndoManager and restore it on undo

alecgeatches committed:
Temporary commit to figure out types

alecgeatches committed:
Fix store type

chriszarate committed:
Warehouse needed types in block-editor typings

alecgeatches committed:
Provide selection-related handlers from data package to sync

alecgeatches committed:
Add BlockSelectionHistory class to track prior valid selections

alecgeatches committed:
Use data-provided methods to track and set selection changes. Have `backupPositions` saved but not yet implemented

alecgeatches committed:
Restore from backupPositions, and include full-block selections without an offset in selection history

alecgeatches committed:
Update BlockSelectionHistory to use relative positions to avoid findBlockByClientIdInDoc() calls

alecgeatches committed:
Add BlockSelectionHistory tests

alecgeatches committed:
Update BlockSelectionHistory to save the current and last selection in the current block to properly restore cursor position in same-block undos

alecgeatches committed:
Fix off-by-one errors when redoing an action

alecgeatches committed:
Simplify BlockSelectionHistory by removing getLastSelection(). Processing selection updates after a setTimeout() allows relative positions to be created correctly with the updated underlying Y.Text data

alecgeatches committed:
Compare selections property-by-property instead of with JSON.stringify()

alecgeatches committed:
Remove additional UndoManager console logging

alecgeatches committed:
Remove unnecessary cross-package references

alecgeatches committed:
Use a map of selection histories to ydoc instances instead of just using one history

alecgeatches committed:
Minor BlockSelectionHistory refactor

alecgeatches committed:
Remove unused BlockSelectionHistory test

alecgeatches committed:
Remove unused sync package typings and tsconfig changes

alecgeatches committed:
Use resetSelection() to allow restoring start and end selection

alecgeatches committed:
Fix resolver name update

alecgeatches committed:
Fix getEntityRecord default handlers in tests

alecgeatches commented on packages/sync/src/utils.ts:
```diff
@@ -82,3 +83,54 @@ export function deserializeCrdtDoc(
 	}
 }
+
+// Convenience types to manage block values with a clientId, attributes, and innerBlocks.
```
@chriszarate I ended up copying over some of the convenience types from our plugin to implement `findBlockByClientIdInDoc()`. I know you had some recent type improvement changes for `Y.Map`, could  those be useful here as well? Can we share them in a sane way?

chriszarate reviewed (commented):
My initial question is: Why can’t this be handled by `getChangesFrom` / `applyChangesToCRDTDoc` (with some refactoring)? It’s very attractive to keep those functions in full control of the sync behavior. Splitting the logic brings complications and hidden dependencies between the packages. 

Even though we don’t sync `selection`, could those mentioned functions read and mutate the selection in the way you’ve implemented here? Could `applyChangesToCRDTDoc` provide some selection hints that can be read by a naïve undo manager? I’m reading this PR quickly on mobile, so I’m sure I’m missing something.

alecgeatches committed:
Pass change origin to getPostChangesFromCRDTDoc() for undo detection, disable yUndoManager event callbacks

alecgeatches committed:
Pass missing origin in getChangesFromCRDTDoc()

alecgeatches committed:
Track selection history in applyPostChangesToCRDTDoc() with BlockSelectionHistory class

alecgeatches committed:
Attempt undo meta management through applyPostChangesToCRDTDoc + editEntityRecord

alecgeatches committed:
Change BlockSelectionHistory to store relative position start and end, attempt to restore selection from core data.

alecgeatches committed:
Restore start and end selections from BlockSelectionHistory. Still doesn't work with cross-block selections when one block was deleted due to order of operations.

alecgeatches committed:
Cheat and use resetSelection in cross-block selection restoration

alecgeatches committed:
Selectively restore selection via editRecord or deferred with resetSelection depending on blocks available

alecgeatches committed:
Move BlockSelectionHistory tests to core-data, update to work with stored start + end selections

alecgeatches committed:
Fix tests and remove unused editEntityRecord handler

alecgeatches committed:
Remove unused code in sync package

alecgeatches committed:
Comment additions and whitespace cleanup

alecgeatches committed:
Fix origin parameter in getChangesFromCRDTDoc() tests

alecgeatches commented on packages/core-data/src/entities.js:
```diff
@@ -360,13 +360,15 @@ async function loadPostTypeEntities() {
 					 *
 					 * @param {import('@wordpress/sync').CRDTDoc}    crdtDoc
 					 * @param {import('@wordpress/sync').ObjectData} editedRecord
+					 * @param {import('@wordpress/sync').Origin}     origin
```
@chriszarate I ended up not actually using this `origin` parameter, newly piped into `getChangesFromCRDTDoc` from `updateEntityRecord()`. I expected to need it when determining if a change came from undo manager versus elsewhere, but the eventual strategy used here didn't need it. I'm keeping it in for now because it wasn't too hard to include this information, and I could see the origin of changes in `getPostChangesFromCRDTDoc()` being useful in the future.

Happy to remove the origin as a parameter if you disagree.

chriszarate replied:
Let's remove if it's unused. Can always add back when needed

alecgeatches replied:
Fixed in https://github.com/WordPress/gutenberg/pull/73066/commits/fa08953f2ee4354708b92822a3ff00d50e3af885 and https://github.com/WordPress/gutenberg/pull/73066/commits/fa08953f2ee4354708b92822a3ff00d50e3af885!

alecgeatches commented on packages/core-data/src/utils/selection.ts:
```diff
@@ -0,0 +1,152 @@
+			if ( isSelectionInSameBlock ) {
+				// Case 3: After content is restored, the selection is available
+				// within the same block
+				resetSelection( selectionStart, selectionEnd, null );
```
I still ended up needing to call `resetSelection()` here anyway, as `editRecord()` doesn't have any effect on visual selection when called deferred. I'm not sure how to get around this. `editRecord()` only changes the selection because of a call to `resetSelection()` in [`useBlockSync()`](https://github.com/Automattic/gutenberg/blob/b3d75ad10705113c75f7979e7de36ad1910e3ef4/packages/block-editor/src/components/provider/use-block-sync.js#L188-L194) anyway, which is [triggered only on block changes](https://github.com/Automattic/gutenberg/blob/b3d75ad10705113c75f7979e7de36ad1910e3ef4/packages/block-editor/src/components/provider/use-block-sync.js#L196).

Our only options here seem to be to call `resetSelection()` ourselves, or send some kind of no-op change that causes `controlledBlocks` to be modified in `editRecord()` in order to trigger a selection change indirectly.

alecgeatches replied:
@chriszarate Based on your feedback I've removed the use of `editRecord()` entirely in https://github.com/WordPress/gutenberg/pull/73066/commits/d2e9f8525d4f62fa151181786852f321850d9f14. We just always wait for the event loop to finish and then use `resetSelection()` instead of sometimes using one or the other.

alecgeatches commented on packages/core-data/src/utils/selection.ts:
```diff
@@ -0,0 +1,152 @@
+ * WordPress dependencies
+ */
+import { dispatch } from '@wordpress/data';
+// @ts-expect-error No exported types.
```
Better way to handle this?

chriszarate commented on packages/sync/src/types.ts:
```diff
@@ -64,11 +64,25 @@ export interface CollectionHandlers {
+	selectionEnd: WPBlockSelection;
+	selectionStart: WPBlockSelection;
+}
+
```
these seem unused

alecgeatches replied:
Good catch, removed in https://github.com/WordPress/gutenberg/pull/73066/commits/9d12d7a91fdb74736e763ed1d74b94348a2af934.

chriszarate commented on packages/sync/src/types.ts:
```diff
@@ -64,11 +64,25 @@ export interface CollectionHandlers {
 	saveRecord: () => Promise< void >;
+
+	addUndoMeta?: ( ydoc: Y.Doc, meta: Map< string, any > ) => void;
+	restoreUndoMeta?: ( ydoc: Y.Doc, meta: Map< string, any > ) => void;
```
nitpick: can you preserve alpha order when it is present?

alecgeatches replied:
Changed these to be alphabetical in types in https://github.com/WordPress/gutenberg/pull/73066/commits/1ad8dd1a943ae6328ccb70262b2916c19123768f.

chriszarate commented on packages/sync/src/types.ts:
```diff
@@ -112,5 +127,11 @@ export interface SyncManager {
+		handlers: {
+			addUndoMeta?: RecordHandlers[ 'addUndoMeta' ];
+			restoreUndoMeta?: RecordHandlers[ 'restoreUndoMeta' ];
+		}
```
```suggestion
		handlers: Pick< RecordHandlers, 'addUndoMeta' | 'restoreUndoMeta' >
```

chriszarate commented on packages/sync/src/undo-manager.ts:
```diff
@@ -50,10 +61,36 @@ export function createUndoManager(): SyncUndoManager {
+			handlers: {
+				addUndoMeta?: RecordHandlers[ 'addUndoMeta' ];
+				restoreUndoMeta?: RecordHandlers[ 'restoreUndoMeta' ];
+			}
```
```suggestion
			handlers: Pick< RecordHandlers, 'addUndoMeta' | 'restoreUndoMeta' >
```

alecgeatches replied:
Fixed both of these instances in https://github.com/WordPress/gutenberg/pull/73066/commits/1ad8dd1a943ae6328ccb70262b2916c19123768f.

chriszarate commented on packages/sync/src/undo-manager.ts:
```diff
@@ -22,9 +30,12 @@ import type { ObjectData, SyncUndoManager } from './types';
-		// Throttle undo/redo captures. (default: 500ms)
-		captureTimeout: 200,
+		// Throttle undo/redo captures.
+		captureTimeout: 500,
```
Any particular reason for this (aside from aligning with comment)? If there's rationale would be good to put in comment.

alecgeatches replied:
500ms experimentally felt the best to me. 200 was a little quick, so sometimes undo stack items would be created mid-word if I slowed while typing too much. 500 felt like a better approximation for the amount of time spent not typing to divide a unit of undo. I'm not sure how I'd document this other than:

```js
captureTimeout: 500 // feels right
```

and the prior choice was also fairly arbitrary. Any idea how to better document this given my subjective methodology?

chriszarate replied:
I think just documenting that it was based on limited subjective local UX testing is good enough. Some context to anchor future potential improvements.

alecgeatches replied:
Good idea, added in https://github.com/WordPress/gutenberg/pull/73066/commits/e31a04900266a5a5ccba8c6097becd214ef1314e:

https://github.com/WordPress/gutenberg/blob/e31a04900266a5a5ccba8c6097becd214ef1314e/packages/sync/src/undo-manager.ts#L36-L40

chriszarate commented on packages/sync/src/undo-manager.ts:
```diff
@@ -50,10 +61,36 @@ export function createUndoManager(): SyncUndoManager {
+				// Necessary for a type check, but this shouldn't happen.
+				return;
+			}
+
```
Move guard to top of method?

alecgeatches replied:
Fixed in https://github.com/WordPress/gutenberg/pull/73066/commits/b2f66081b64814e01a390b0ecf649cb169a06a23.

chriszarate commented on packages/sync/src/undo-manager.ts:
```diff
@@ -50,10 +61,36 @@ export function createUndoManager(): SyncUndoManager {
+			const { addUndoMeta, restoreUndoMeta } = handlers;
+
+			yUndoManager.on( 'stack-item-added', ( event: StackItemEvent ) => {
+				addUndoMeta?.( ydoc, event.stackItem.meta );
```
Why would this handler be undefined?

alecgeatches replied:
No reason. Fixed in https://github.com/WordPress/gutenberg/pull/73066/commits/1ad8dd1a943ae6328ccb70262b2916c19123768f.

alecgeatches committed:
Remove unused origin param from getChangesFromCRDTDoc()

alecgeatches committed:
Remove origin parameter from crdt tests

alecgeatches committed:
Avoid passing editRecord() to restoreSelection, always wait for event loop before restoring selection

alecgeatches committed:
Cleanup undoMeta handlers - alphabetize, make required, and use Pick types

alecgeatches committed:
Put null check at top of function

alecgeatches committed:
Remove unused sync types

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+
+		const attributes = block?.get( 'attributes' ) as
+			| Y.Map< Y.Text >
+			| undefined;
```
Assertion not necessary with `YMapWrap` :)
```suggestion
		const attributes = block?.get( 'attributes' );
```

alecgeatches replied:
Merged in with your changes in https://github.com/WordPress/gutenberg/pull/73066/commits/bc818f3b8076708c8d160908f9f4cf3c0a022146.

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	/**
+	 * Convert a WPSelection to a Position (relative position or block position).
```
```suggestion
	 * Convert a WPBlockSelection to a Position (relative position or block position).
```

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+	/**
+	 * Convert a WPSelection to a Position (relative position or block position).
+	 * @param selection
+	 * @return Position or null if conversion fails
```
this return annotation is inaccurate, no null return
```suggestion
```

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+
+	/**
+	 * Add a position to the history, maintaining only the last `historySize` unique selections.
+	 * New positions are added to the front.
```
These are selections, not positions, right? Or am I missing some nuance?
```suggestion
	 * Add a selection to the history, maintaining only the last `historySize` unique selections.
	 * New selections are added to the front.
```

alecgeatches replied:
That's correct, these should all be called "selections". Some back-and-forth on naming during development. I'll do a find for "position" used incorrectly here and make sure they reflect the correct name.

chriszarate committed:
Simplify block selection logic

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+ * - history: An array of previous blocks. Use this as a backup for restoration
+ *            locations.
+ */
+export class BlockSelectionHistory {
```
consider factory function with state in closure

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,210 @@
+	 * Get the block history (previous blocks only, not current or last selection).
+	 * @param count Number of positions to retrieve
+	 */
+	public getSelectionHistory( count: number ): YFullSelection[] {
```
does passing a count have a use case? currently we import and pass the default? can we remove for simplicity?

alecgeatches replied:
Fully removed in your commit https://github.com/WordPress/gutenberg/pull/73066/commits/bc818f3b8076708c8d160908f9f4cf3c0a022146 + some fixes in https://github.com/WordPress/gutenberg/pull/73066/commits/5cffe7c38554037a54a43e12434744b83d3e633f.

alecgeatches committed:
Add explanation for captureTimeout magic number

alecgeatches committed:
Refactor BlockSelectionHistory:

- Remove count from getSelectionHistory(), replace existing calls
- Remove use of "position" and replace with "selection" in comments
- Update incorrect comments

alecgeatches committed:
Update BlockSelectionHistory to use factory function instead of class.

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,184 @@
+}
+
+/**
+ * This class is used to track recent block selections to help in restoring
```
```suggestion
 * This function is used to track recent block selections to help in restoring
```

chriszarate commented on packages/core-data/src/utils/block-selection-history.ts:
```diff
@@ -0,0 +1,184 @@
+	/**
+	 * Get the current selection (most recent selection in the current block).
+	 */
+	const getCurrentSelection = () => {
```
Would you mind adding return type annotations to all functions in this file? Most are covered by the `BlockSelectionHistory` type, but I think it's always good practice to avoid inferred return types. It helps prevent unintentional changes to the return type, which can lead to type bugs in consuming code.

chriszarate commented on packages/core-data/src/utils/crdt.ts:
```diff
@@ -31,14 +31,24 @@ import {
+import {
+	createBlockSelectionHistory,
+	YSelectionType,
+} from './block-selection-history';
```
```suggestion
import {
	createBlockSelectionHistory,
	YSelectionType,
	type BlockSelectionHistory,
	type YFullSelection,
	type YSelection,
	type YSelectionHistory,
} from './block-selection-history';
```

alecgeatches committed:
Re-add block selection history function types, reorder imports, fix a comment

chriszarate committed:
Keep selection history as single array with current selection at 0

chriszarate committed:
Update tests for block selection history changes

chriszarate committed:
Rename util selection.ts => crdt-selection.ts

chriszarate committed:
Remove unused equality functions

chriszarate committed:
Reorganize CRDT selection functions

chriszarate reviewed (approved):
I added a few (hopefully) simplifying commits. This looks great!

alecgeatches committed:
Fix selection function import
